neotron-bmc-pico/src/main.rs

```diff
@@ -510,6 +526,46 @@ mod app {
 					panic!("Q full!");
 				}
 			}
+
+			// The speaker PWM needs to be updated (register was updated)
+			if register_state.speaker_needs_update {
+				defmt::info!("speaker PWM update");
+				register_state.speaker_needs_update = false;
+
+				let config = &mut register_state.speaker_config;
+				let duration = config.duration();
+
+				if duration > 0 {
+					// a note has to be played - enable PWM
+					ctx.shared.speaker_pwm_tim.lock(|tim14| {
```

Comment:
I think this code that pokes TIM14 could be moved into the `pwm` (or `speaker`) module. Something like:

```rust
register_state.speaker.update();
```

The boolean flag could also be hidden inside the `Speaker` object:

```rust
if register_state.speaker.needs_update() {
    register_state.speaker.update();
}
```

Reply:
I've done something which resembles that, but I guess for it to be exactly like  that, the struct would need to hold both the register state and the hardware. Unless I'm misunderstanding.